### ISSUE TITLE
Implicit calls to `deref` and `deref_mut` must be forbidden in raw pointer operations.

### ISSUE THREAD

**Issue by AngelicosPhosphoros** [rust-lang/rust#131847]
I tried this code:

```rust
use std::ops::Deref;
use std::mem::MaybeUninit;

struct A{
    b: MaybeUninit<B>,
}

struct B{
    field: String,
}

impl Deref for A {
    type Target = B;
    fn deref(&self)->&B{
        println!("Called deref!");
        // SAFETY: Valid instances of A have b initialized.
        unsafe {
            self.b.assume_init_ref()
        }
    }
}

fn main() {
    let mut a: MaybeUninit<A> = MaybeUninit::uninit();
    unsafe {
        let p = a.as_ptr();
        // The only way to get pointer to field of pointee.
        let b_ptr = &raw const (*p).b;
        // However, it may cause calls to deref if we are not vigilant.
        let field_ptr = &raw const (*p).field;
    }
}
```

I expected to see this happen: this code should be rejected because it causes implicit calls to `Deref::deref` which assumes `*p` to be initialized.

Instead, this happened: code compiles and prints `Called deref!` when executed.

### Meta

`rustc --version --verbose`:
```
rustc 1.84.0-nightly (9322d183f 2024-10-14)
binary: rustc
commit-hash: 9322d183f45e0fd5a509820874cc5ff27744a479
commit-date: 2024-10-14
host: x86_64-unknown-linux-gnu
release: 1.84.0-nightly
LLVM version: 19.1.1
```

**Comment by AngelicosPhosphoros** [comment]
I think, it should call derefs only this way:
```
(**p) <-- Explicitly dereference object again.
(*).deref() <-- Explicitly call deref method.
```

**Comment by carbotaniuman** [comment]
While an unfortunate footgun, this is not a soundness issue. Removing the `unsafe` shows that only the first line is (correctly) exempt from needing an unsafe block, while the second line does require one.

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7545201409c69d43e6b2d9700b04aef3

**Comment by VorpalBlade** [comment]
It would make sense to have a lint for this still (clippy or rustc).

**Comment by fmease** [comment]
Might get caught by https://github.com/rust-lang/rust/pull/123239, haven't checked.

**Comment by compiler-errors** [comment]
This is not unsound, so I'm gonna untag it as such.

**Comment by traviscross** [comment]
@rustbot labels -C-bug +C-discussion

Note that this isn't specific to raw references.  It's just as plausible to do this with normal references, e.g.:

```rust
use core::{mem::MaybeUninit, ops::Deref};

struct A(MaybeUninit<B>);
struct B {
    field: String,
}

impl Deref for A {
    type Target = B;
    fn deref(&self) -> &B {
        unsafe { self.0.assume_init_ref() }
    }
}

fn main() {
    let a: MaybeUninit<A> = MaybeUninit::uninit();
    let p = a.as_ptr();
    let bad = unsafe { &(*p).field };
    println!("{}", bad);
}
```

**Comment by RalfJung** [comment]
I would argue you asked for trouble by writing that `Deref` instance. That instance is unsound, it can be used to cause UB in safe code:
```rust
fn main() {
    let a = A { b: MaybeUninit::uninit() };
    let val = &a.field;
    println!("{val}");
}
```

So, the fix is to not write such `Deref` instances. I am not convinced a lint is justified. (Well, ideally we'd have a lint against unsound code, but alas, that's not possible. ;)

**Comment by RalfJung** [comment]
> because it causes implicit calls to Deref::deref which assumes *p to be initialized.

`deref` is not allowed to make such assumptions. It is a safe function, and as such must not make any assumptions beyond being given a well-typed argument.

Cc @rust-lang/opsem

**Comment by ChayimFriedman2** [comment]
@RalfJung The problem (AFAIK) is not that Rust is unsound. Rust is behaving correctly. The problem is that it's too easy to make mistakes.

Making a safe `Deref` is fine if the type has an invariant that it's always initialized, but private code may break this invariant temporarily, and it is too easy to invoke `Deref` by mistake then. Ultimately, we should be careful when writing unsafe code, and the language could assist with that.

**Comment by RalfJung** [comment]
Yes we use minimal examples, but this was too minimal. ;) The safety invariant and abstraction barrier are a key part of the argument here. Given there were no comments to indicate anything like that, it is reasonable to assume there is no such invariant.

Thanks for clarifying!

**Comment by AngelicosPhosphoros** [comment]
I added comments.
